src/vs/workbench/contrib/files/browser/views/explorerViewer.ts

```diff
@@ -980,12 +980,21 @@ export class FileDragAndDrop implements ITreeDragAndDrop<ExplorerItem> {
 
 		try {
 
-			// Desktop DND (Import file)
+			// External File DND (Import file)
 			if (data instanceof NativeDragAndDropData) {
-				if (isWeb) {
-					const browserUpload = this.instantiationService.createInstance(BrowserFileUpload);
-					await browserUpload.upload(target, originalEvent);
-				} else {
+				// Native DND from Desktop
+				if (originalEvent.dataTransfer?.types.indexOf('Files') !== -1) {
+					if (isWeb) {
+						const browserUpload = this.instantiationService.createInstance(BrowserFileUpload);
+						await browserUpload.upload(target, originalEvent);
+					} else {
+						const nativeImport = this.instantiationService.createInstance(NativeFileImport);
```

Comment:
Code duplication, the exact same lines are duplicate below.

Reply:
I felt the 2 lines duplicated were more clear in terms of readability (understanding the code paths), but I was mistaken.

Addressing this issue and the one below about browserfileupload:

My understanding is that BrowserFileUpload is intended for uploading files from the local file system. This means that the chrome has done the work of populating the data transfer object, not us. So when Files exists, we know that it is a file on the local machine and we have 2 handlers for it. One for native and one for web.

On the other hand, when dragging FS-provided files, the DTO is populated by us in the resourceURLs. The nativefileimport handles this in desktop and inweb exactly the same way because it hands it off to the file system provider.

Perhaps some renaming of these helper classes is in order?

Reply:
Yeah thanks for clarifying, fully agree that a rename would help if both can run in web and desktop irrespectively.